### PR TITLE
SIDM-4233 SSO Login

### PR DESCRIPTION
**Pull request by nikola-naydenov-hmcts**
Complete federation login flow

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -64,8 +64,18 @@ allprojects {
...
     implementation group: 'org.springframework.boot', name: 'spring-boot-starter-security'
-    // TODO: remove version once 2.2.2.RELEASE is out
-    implementation(group: 'org.springframework.cloud', name: 'spring-cloud-starter-netflix-zuul', version: '2.2.1.RELEASE') {
+
+    implementation group: 'org.springframework.security', name: 'spring-security-web'
+    implementation group: 'org.springframework.security', name: 'spring-security-data'
+    implementation group: 'org.springframework.security', name: 'spring-security-config'
+    implementation group: 'org.springframework.boot', name: 'spring-boot-starter-oauth2-client'
+    implementation group: 'org.springframework.boot', name: 'spring-boot-starter-oauth2-resource-server'
+    implementation group: 'org.springframework.boot', name: 'spring-boot-starter-data-redis-reactive'
```
Comment (dfourn) [inline build.gradle:73]: So you're gonna use redis for this after all? 
👍 for using the reactive one

File: src/main/java/uk/gov/hmcts/reform/idam/web/sso/SSOAuthenticationSuccessHandler.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+        if (response.isCommitted()) {
+            throw restException(null, HttpStatus.INTERNAL_SERVER_ERROR, new HttpHeaders(),
+                HttpStatus.INTERNAL_SERVER_ERROR.getReasonPhrase(),
+                "Response has already been committed. Unable to redirect.");
+        }
+
+        clearAuthenticationAttributes(request);
+
+        if (!responseUrl.contains("error")) {
```
Comment (dfourn) [inline src/main/java/uk/gov/hmcts/reform/idam/web/sso/SSOAuthenticationSuccessHandler.java:101]: seems like a weird way of checking if the request was successful

File: src/main/java/uk/gov/hmcts/reform/idam/web/Application.java [hunk 3]
```diff
@@ -8,6 +8,7 @@
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
 import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
 import org.springframework.cloud.netflix.zuul.EnableZuulProxy;
+import org.springframework.cloud.openfeign.EnableFeignClients;
```
Comment (tbamido) [inline src/main/java/uk/gov/hmcts/reform/idam/web/Application.java:11]: don't think we are using this?
```suggestion
```

File: src/main/java/uk/gov/hmcts/reform/idam/web/helper/ErrorHelper.java [hunk 4]
```diff
@@ -27,4 +36,25 @@ public static void showError(String errorTitle, String errorMessage, String erro
...
+            return HttpClientErrorException.create(message, status, status.getReasonPhrase(), headers, body, UTF_8);
+        }
+        return HttpServerErrorException.create(message, status, status.getReasonPhrase(), headers, body, UTF_8);
+    }
+
+    public static String jsonErrorString(String error, String description) {
+        return "{\n" +
+            "  \"error\": \"errorString\",\n" +
+            "  \"description\": \"errorDescription\"\n" +
+            "}";
```
Comment (tbamido) [inline src/main/java/uk/gov/hmcts/reform/idam/web/helper/ErrorHelper.java:58]: Should be returning values from eror and description variables:
```suggestion
        return String.format("{%n \"error\": \"%s\",%n  \"description\": \"%s\"%n}", error, description);
```

File: src/main/java/uk/gov/hmcts/reform/idam/web/sso/SSOAuthenticationSuccessHandler.java (hunk 2 above)
Comment (tbamido) [reply src/main/java/uk/gov/hmcts/reform/idam/web/sso/SSOAuthenticationSuccessHandler.java:101]: @dfourn this request will return a 302 with potential error code on it

File: src/main/java/uk/gov/hmcts/reform/idam/web/helper/ErrorHelper.java (hunk 4 above)
Comment (nikola-naydenov-hmcts) [reply src/main/java/uk/gov/hmcts/reform/idam/web/helper/ErrorHelper.java:58]: lol that was the idea but im dumb

**Comment by tbamido** [comment]
should we merge this into the feature/sso branch instead?

**Pull request merged**